observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingChannelInterceptor.java

```diff
@@ -16,33 +16,248 @@
+/**
+ * A logging interceptor for {@code LoggingChannelProvider}.
+ */
 @Internal
 public final class InternalLoggingChannelInterceptor implements ClientInterceptor {
+  private static final Logger logger = Logger
+      .getLogger(InternalLoggingChannelInterceptor.class.getName());
+
+  private final LogHelper helper;
 
   public interface Factory {
     ClientInterceptor create();
   }
 
   public static class FactoryImpl implements Factory {
+    private final Sink sink;
+    private final LogHelper helper;
+
+    static LogHelper createLogHelper(Sink sink, TimeProvider provider) {
+      return new LogHelper(sink, provider);
+    }
+
+    public FactoryImpl(Sink sink) {
+      this.sink = sink;
+      this.helper = createLogHelper(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
+    }
 
     @Override
     public ClientInterceptor create() {
-      return new InternalLoggingChannelInterceptor();
+      return new InternalLoggingChannelInterceptor(helper);
+    }
+
+    public void close() {
+      sink.close();
     }
   }
 
+  private InternalLoggingChannelInterceptor(LogHelper helper) {
+    this.helper = helper;
+  }
+
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
       CallOptions callOptions, Channel next) {
-    // TODO(dnvindhya) implement the interceptor
-    return null;
+
+    final AtomicLong seq = new AtomicLong(1);
+    final String rpcId = UUID.randomUUID().toString();
+    final String authority = next.authority();
+    final String serviceName = method.getServiceName();
+    final String methodName = method.getBareMethodName();
+    // Get the stricter deadline to calculate the timeout once the call starts
+    final Deadline deadline = LogHelper.min(callOptions.getDeadline(),
+        Context.current().getDeadline());
+
+    // TODO (dnvindhya): implement isMethodToBeLogged() to check for methods to be logged
+    // according to config. Until then always return true.
+    if (!helper.isMethodToBeLogged(method.getFullMethodName())) {
+      return next.newCall(method, callOptions);
+    }
+
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        // Event: EventType.GRPC_CALL_REQUEST_HEADER
+        // The timeout should reflect the time remaining when the call is started, so compute
+        // remaining time here.
+        final Duration timeout = deadline == null ? null
+            : Durations.fromNanos(deadline.timeRemaining(TimeUnit.NANOSECONDS));
+
+        try {
+          helper.logRequestHeader(
+              seq.getAndIncrement(),
+              serviceName,
+              methodName,
+              authority,
+              timeout,
+              headers,
+              EventLogger.LOGGER_CLIENT,
+              rpcId,
+              null);
+        } catch (IllegalArgumentException e1) {
```

Comment:
You are doing the exact same thing in all 3 catch clauses. And the last one (`Exception`) is the super class of the other 2. So you can just remove the first 2 catch clauses.

Reply:
Intention was to catch expected exceptions instead of super class, and the ones which does not fall in expected will be caught as part of Generic exception.

Updated to catch generic exception, and added comment saying the exceptions expected.